src/Lykke.Service.CryptoIndex.DomainServices/LCI10/LCI10Calculator.cs

```diff
@@ -0,0 +1,300 @@
+            }
+
+            _log.Info($"Finished calculating weights for {assets.Count} assets.");
+        }
+
+        private async Task CalculateIndex(ITimerTrigger timer, TimerTriggeredHandlerArgs args, CancellationToken ct)
+        {
+            try
+            {
+                await CalculateIndex();
+            }
+            catch (Exception e)
+            {
+                _log.Error(e);
+            }
+        }
+
+        private async Task CalculateIndex()
+        {
+            _log.Info("Started calculating index...");
+
+            var settings = await _settingsService.GetAsync();
+
+            List<AssetMarketCap> marketCaps;
+            IDictionary<string, decimal> weights;
+            var assetsPrices = await _tickPricesService.GetPricesAsync();
+            var lastIndex = await _indexStateRepository.GetAsync();
+
+            RecalculateTheWeightsIfSomeWeightsAreNotFound(settings.Assets);
+
+            lock (_sync)
+            {
+                marketCaps = _marketCaps.ToList();
+                weights = _weights.Clone();
+            }
+
+            if (!IsAllDataArePresented(marketCaps, weights, assetsPrices))
+            {
+                _log.Info("Skipped LCI10 calculation.");
+                return;
+            }
+
+            // Save index as previous for the next execution
+            var indexState = CalculateIndexState(settings.Assets, weights, assetsPrices, lastIndex);
+            await _indexStateRepository.SetAsync(indexState);
+
+            // Save index to history
+            var indexHistory = new IndexHistory(indexState.Value, marketCaps, weights, indexState.MiddlePrices, DateTime.UtcNow);
+            await _indexHistoryRepository.InsertAsync(indexHistory);
+
+            // Publish index to RabbitMq
+            var tickPrice = new Domain.TickPrice.TickPrice(Lci10, Lci10, indexHistory.Value, indexHistory.Value, indexHistory.Time);
+            _tickPricePublisher.Publish(tickPrice);
+
+            _log.Info($"Finished calculating index for {settings.Assets.Count} assets, value: {indexState.Value}.");
+        }
+
+        private IndexState CalculateIndexState(IReadOnlyCollection<string> assets, IDictionary<string, decimal> assetsWeights,
+            IDictionary<string, IDictionary<string, decimal>> assetsPrices, IndexState lastIndex)
+        {
+            var middlePrices = GetMiddlePrices(assetsPrices);
+
+            if (lastIndex == null)
+                lastIndex = new IndexState(1000, middlePrices);
+
+            CheckThatAllAssetsArePresent(assets, assetsWeights, assetsPrices);
+
+            var rLci10 = 0m;
+
+            foreach (var asset in assets)
+            {
+                var middlePrice = GetMiddlePrice(asset, assetsPrices);
+                var previousMiddlePrice = GetPreviousMiddlePrice(asset, lastIndex.MiddlePrices, middlePrice);
+
+                var weight = assetsWeights[asset];
+
+                var r = middlePrice / previousMiddlePrice;
```

Comment:
Please, rename the variable "r"

Reply:
It is "R" in the specification, don't know a better name.